File: platforms/ios/Dezel/Dezel/View/Transition/TransitionManager.swift
/**
 * @class TransitionManager
 * @since 0.7.0
 */
public final class TransitionManager {

	//--------------------------------------------------------------------------
	// MARK: Types
	//--------------------------------------------------------------------------

	/**
	 * @type Callback
	 * @since 0.7.0
	 */
	public typealias Callback = (() -> Void)

	//--------------------------------------------------------------------------
	// MARK: Static
	//--------------------------------------------------------------------------

	/**
	 * @property transition
	 * @since 0.7.0
	 */
	public static var transition: Transition? {
		return transitions.last
	}

	/**
	 * @method begin
	 * @since 0.7.0
	 */
	public static func begin(duration: CFTimeInterval, equation: CAMediaTimingFunction, delay: CFTimeInterval, callback: @escaping Callback) {

		let transition = Transition()

		Synchronizer.main.execute()

		transition.begin()
		transition.delay = delay
		transition.duration = duration
		transition.equation = equation
		transition.callback = {
			transition.callback = nil
			transition.finished()
			transition.reset()
			callback()
		}

		TransitionManager.transitions.append(transition)
	}

	/**
	 * @method commit
	 * @since 0.7.0
	 */
	public static func commit() {

		guard let transition = TransitionManager.transition else {
			return
		}

		Synchronizer.main.execute()

		transition.commit()

		TransitionManager.transitions.removeLast()
	}

	//--------------------------------------------------------------------------
	// MARK: Private API
	//--------------------------------------------------------------------------

	/**
	 * @property transitions
	 * @since 0.7.0
	 * @hidden
	 */
	private static var transitions: [Transition] = []
}

File: platforms/android/Dezel/dezel/src/main/java/ca/logaritm/dezel/view/animation/TransitionManager.kt
package ca.logaritm.dezel.view.transition

import android.view.animation.Interpolator
import ca.logaritm.dezel.application.ApplicationActivity
import ca.logaritm.dezel.extension.type.last
import ca.logaritm.dezel.view.Synchronizer

/**
 * @object TransitionManager
 * @since 0.7.0
 * @hidden
 */
public object TransitionManager {

	//--------------------------------------------------------------------------
	// Properties
	//--------------------------------------------------------------------------

	/**
	 * @property transition
	 * @since 0.7.0
	 */
	public val transition: Transition?
		get() = transitions.last

	//--------------------------------------------------------------------------
	// Methods
	//--------------------------------------------------------------------------

	/**
	 * @method begin
	 * @since 0.7.0
	 */
	public fun begin(activity: ApplicationActivity, duration: Double, equation: Interpolator, delay: Double, callback: TransitionCallback) {

		val transition = Transition(activity)

		Synchronizer.main.execute()

		transition.callback = callback
		transition.duration = duration
		transition.equation = equation
		transition.delay = delay
		transition.begin()

		transitions.add(transition)
	}

	/**
	 * @method commit
	 * @since 0.7.0
	 */
	public fun commit() {

		val transition = transitions.last
		if (transition == null) {
			return
		}

		Synchronizer.main.execute()

		transition.commit()

		transitions.remove(transition)
	}

	//--------------------------------------------------------------------------
	// Private API
	//--------------------------------------------------------------------------

	/**
	 * @property transitions
	 * @since 0.7.0
	 * @hidden
	 */
	private var transitions: MutableList<Transition> = mutableListOf()
}
